ddtrace/propagation/http.py

```diff
@@ -483,10 +492,214 @@ def _extract(headers):
+
+        4. If the traceparent sampled flag == 0, and the tracestate sampling priority is less than or equal to 0,
+        use the extracted sampling priority value.
+        """
+
+        if sampling_priority_tp == 0 and (not sampling_priority_ts or sampling_priority_ts >= 0):
+            sampling_priority = 0
+
+        elif sampling_priority_tp == 1 and (not sampling_priority_ts or sampling_priority_ts < 0):
+            sampling_priority = 1
+
+        elif sampling_priority_tp == 1 and sampling_priority_ts > 0:
+            sampling_priority = sampling_priority_ts
+
+        elif sampling_priority_tp == 0 and sampling_priority_ts < 0:
+            sampling_priority = sampling_priority_ts
+
+        return sampling_priority
+
+    @staticmethod
+    def _extract(headers):
+        # type: (Dict[str, str]) -> Optional[Context]
+
+        try:
+            tp = _extract_header_value(_POSSIBLE_HTTP_HEADER_TRACEPARENT, headers)
+            if tp is None:
+                return None
+            traceparent_values = _TraceContext._get_traceparent_values(tp)
+        except (ValueError, AssertionError):
+            log.exception("received invalid w3c traceparent: %s ", tp)
+            return None
+
+        trace_id, span_id, sampling_priority = traceparent_values
+        meta = {"traceparent": tp}
+
+        ts = _extract_header_value(_POSSIBLE_HTTP_HEADER_TRACESTATE, headers)
+        if ts:
+            # the value MUST contain only ASCII characters in the
+            # range of 0x20 to 0x7E except comma (,) and equal-sign (=)
+            if re.search(r"[^\x20-\x7E]+", ts):
+                log.debug("received invalid tracestate header: %r", ts)
+            else:
+                # store ts so we keep other vendor data, even if dd ends up being invalid
+                meta[_TRACESTATE_KEY] = ts
+                try:
+                    tracestate_values = _TraceContext._get_tracestate_values(ts)
+                except (TypeError, ValueError, AttributeError):
+                    log.debug(("received invalid dd header value in tracestate: %r "), ts)
+                    tracestate_values = None
+
+                if tracestate_values:
+                    sampling_priority_ts, other_propagated_tags, origin = tracestate_values
+                    meta.update(other_propagated_tags)
+
+                    sampling_priority = _TraceContext._get_sampling_priority(sampling_priority, sampling_priority_ts)
+
+                    return Context(
+                        trace_id=trace_id,
+                        span_id=span_id,
+                        sampling_priority=sampling_priority,
+                        dd_origin=origin,
+                        meta=cast(_MetaDictType, meta),
+                    )
+
+        return Context(
+            trace_id=trace_id,
+            span_id=span_id,
+            sampling_priority=sampling_priority,
+            meta=cast(_MetaDictType, meta),
+        )
```

Comment:
nit: can we combine these return statements into one. Ex: Set `origin=None` on line 660 and remove the return statement on line 682. Returning one context object might be easier to read OR we can create the context object first and then update the fields as parse the tracestate.